### PR TITLE
 Make IAM influence after IAM display, Add click action handler

### PR DESCRIPTION
**Pull request by Jeasmine**
* Changes Tests to accomplish new functionality
   * Minor improvements changes
   * If the user sends and outcome from the click action handler callback, that outcome should be DIRECT influenced by IAM
   * DIRECT influences by IAM can be from click action handler or click action outcome

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1005)

### CODE REVIEW

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalRemoteParams.java [hunk 1]
```diff
@@ -92,12 +92,12 @@ public String toString() {
    private static int androidParamsRetries = 0;
 
    private static final String OUTCOME_PARAM = "outcomes";
-   private static final String V2_PARAM = "v2_active";
+   private static final String OUTCOMES_V2_SERVICE_PARAM = "outcomes_v2_service_enabled";
```
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalRemoteParams.java:95]: We can remove `outcomes_` from the string as this is under the `outcomes` json key anyway. `v2_enabled` I think is the most clear.
We should keep `OUTCOMES_` in the name so it is clear when we use it though.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java [hunk 2]
```diff
@@ -633,6 +635,160 @@ public void testInAppMessageClickActionOutcomeWithValue() throws Exception {
...
+                    put("name", "my_click_name");
+                }}
+        );
+
+        // Ensure we fire public callback that In-App was clicked.
+        assertEquals(lastAction[0].clickName, "my_click_name");
+    }
+
+    @Test
+    public void testOnMessageActionOccurredOnMessageSendOutcomeV2AfterDismiss() throws Exception {
```
Comment (mikechoch) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java:715]: rename this test maybe?
`testIAMActionSendsOutcome_afterDismiss_usingOutcomesV2` or
`testOnIAMActionSendsOutcome_afterDismiss_usingOutcomesV2`

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java [hunk 3]
```diff
@@ -1041,8 +1041,8 @@ public void complete(OneSignalRemoteParams.Params params) {
             );
             OneSignalPrefs.saveBool(
                OneSignalPrefs.PREFS_ONESIGNAL,
-               OneSignalPrefs.PREFS_OS_OUTCOMES_V2,
-                    params.influenceParams.v2Enabled
+               preferences.getOutcomesV2KeyName(),
+                    params.influenceParams.outcomesV2ServiceEnabled
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:1045]: fix extra tabbing

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/OSOutcomeEventsFactory.java [hunk 4]
```diff
@@ -23,11 +23,22 @@ public OSOutcomeEventsFactory(OSLogger logger, OneSignalAPIClient apiClient, One
     public OSOutcomeEventsRepository getRepository() {
         if (repository == null)
             createRepository();
+        else
+            checkVersionChanged();
         return repository;
     }
 
+    private void checkVersionChanged() {
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/OSOutcomeEventsFactory.java:31]: Rename?
`validateRepositoryVersion` or
`validateOutcomeRepositoryVersion`

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java [hunk 5]
```diff
@@ -633,6 +635,160 @@ public void testInAppMessageClickActionOutcomeWithValue() throws Exception {
...
+            @Override
+            public void inAppMessageClicked(OSInAppMessageAction result) {
+                lastAction[0] = result;
+                // Ensure we are on the main thread when running the callback, since the app developer
+                //   will most likely need to update UI.
+                assertMainThread();
+
+                OneSignal.sendOutcome("test");
+                // Ensure outcome is sent
+                ShadowOneSignalRestClient.Request iamOutcomeRequest = ShadowOneSignalRestClient.requests.get(4);
```
Comment (mikechoch) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java:675]: Below you are doing a lot of `Request` asserts
We should continuously update `RestClientAsserts.java` whenever we find new endpoints for testing
Ask me for help or check the current implementation and try to mimic it for the `measure_sources` endpoint

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java [hunk 6]
```diff
@@ -633,6 +635,160 @@ public void testInAppMessageClickActionOutcomeWithValue() throws Exception {
...
+        );
+
+        // Ensure we fire public callback that In-App was clicked.
+        assertEquals(lastAction[0].clickName, "my_click_name");
+
+        OneSignalPackagePrivateHelper.dismissCurrentMessage();
+
+        OneSignal.sendOutcome("test1");
+        // Ensure outcome is sent but with INDIRECT influence from IAM
+        ShadowOneSignalRestClient.Request iamOutcomeRequest = ShadowOneSignalRestClient.requests.get(5);
```
Comment (mikechoch) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java:770]: Use `RestClientAsserts.java`, similar to above comment

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java [hunk 7]
```diff
@@ -633,6 +622,130 @@ public void testInAppMessageClickActionOutcomeWithValue() throws Exception {
...
+            public void inAppMessageClicked(OSInAppMessageAction result) {
+                lastAction[0] = result;
+                // Ensure we are on the main thread when running the callback, since the app developer
+                //   will most likely need to update UI.
+                assertMainThread();
+
+                OneSignal.sendOutcome("test");
+                try {
+                    // Ensure outcome is sent
+                    assertMeasureOnV2AtIndex(4, "test", new JSONArray().put(message.messageId), new JSONArray(), null, null);
```
Comment (mikechoch) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java:663]: NICE!

**Pull request merged**